pkg/channeld/connection.go

```diff
@@ -423,28 +424,31 @@ func (c *Connection) receive() {
 		packetReceived.WithLabelValues(c.connectionType.String()).Inc()
 
 	*/
-
-	for bufPos := 0; bufPos < c.readPos; {
+	bufPos := 0
+	for bufPos = 0; bufPos < c.readPos; {
 		if c.readPacket(&bufPos) == nil {
-			return
+			break
 		}
 		if bufPos < c.readPos {
 			combinedPacketCount.WithLabelValues(c.connectionType.String()).Inc()
 		}
 	}
 
+	//move unhandled content to the front
+	copy(c.readBuffer, c.readBuffer[bufPos:c.readPos])

```

Comment:
If the tag is invalid, c.readPos will be set to 0, causing out of range error here.

Should add the check of c.IsClosing()

Reply:
you are right, how about remove the c.readPos modification in readPacket function, seems more readable.

Reply:
If you'd like `readPacket` not to change `c.readPos`, I suggest changing its return values to (*Packet, error), and close the connection if err != nil in `receive`.